### DISCUSSION TITLE
Solving a big Delay when use a lookup for more 70,000 rows !

### DISCUSSION THREAD

**Discussion by waleed51** [serenity-is/Serenity#5417] [Help]
I have a (Visits) grid which records visits transactions to my company ,
and it have a relation with (Visitor) -Visit have one or more visitors- ,

I use a Visitors Row as Lookup in Visit dialog, to can select visitor when I add a new Visitor transaction, ( like in Picture).

the problem is Visitor table have more 70,000 rows, when I use it in lookup it make a big delay for searching on it to retrieve data.

Is there is any way to not load all lookup, only load based on search in lookup, or any alternative option compatible with my business?

![Screenshot_3](https://user-images.githubusercontent.com/18492471/106301166-0ed2a380-6260-11eb-9ef7-f6ae3e243bb2.png)

**Comment by kilroyFR** [comment]
There is an option to load lookup asynchronously

**Comment by waleed51** [comment]
Please more explanation @kilroyFR

**Comment by VictorTomaili** [comment]
I guess you want to something like in this demo filters. Some of our editors included in our premium. But you are free to write your own. 
https://serenity.is/demo/AdvancedSamples/DataExplorer

**Reply by waleed51** [reply]
sorry , but i don't know how to do my own?? 
please more help ! it's so critical and important to me!

**Reply by VictorTomaili** [reply]
ServiceLookupEditor is good option what i miss.

**Reply by waleed51** [reply]
how to use it ???

**Comment by minhhungit** [comment]
@waleed51 Use [ServiceLookupEditor], like this
```csharp
[ServiceLookupEditor(service: "Northwind/Customer/List", idField: "CustomerID", textField: "CompanyName", PageSize = 5)]
public String CustomerID { get; set; }
```

**Reply by waleed51** [reply]
first thank you for your help , but it isn't work 
i user it in Columns.cs - Form.cs -  Row.cs  and not work at any one , and give error (ServiceLookupEditor could not be find)

please, i need how to use it ?
@minhhungit 

![image](https://user-images.githubusercontent.com/18492471/106359948-81a25400-631e-11eb-8d6b-9363ea3a8efd.png)

**Reply by minhhungit** [reply]
which serenity version you are using, if you use old version, you might need to implement the editor by yourself,check code here

https://github.com/serenity-is/Serenity/blob/b8fd791681e91014e5ddc7ee24f3efbeb91f0ca0/src/Serenity.Scripts/CoreLib/UI/Editors/ServiceLookupEditor.ts

**Reply by waleed51** [reply]
i use an old version and try to implement the editor like code in link,  but it give a lot of errors .
please help !! 
thank you for ervery thing, and i'm so sorry @minhhungit 

![image](https://user-images.githubusercontent.com/18492471/106370593-1da27e80-6364-11eb-86b7-f5618b38a1a2.png)

![image](https://user-images.githubusercontent.com/18492471/106370599-2a26d700-6364-11eb-8efb-305a633e541b.png)

**Comment by minhhungit** [comment]
@waleed51 use this code, I cloned code from the link https://github.com/serenity-is/Serenity/blob/b8fd791681e91014e5ddc7ee24f3efbeb91f0ca0/src/Serenity.Scripts/CoreLib/UI/Editors/ServiceLookupEditor.ts

```javascript
namespace YOUR_APP.Web.Common.Editors {

    export interface ServiceLookupEditorOptions extends Serenity.Select2EditorOptions {
        service?: string;
        idField: string;
        textField: string;
        pageSize?: number;
        minimumResultsForSearch?: any;
        sort: string[];
        columnSelection?: Serenity.ColumnSelection;
        includeColumns?: string[];
        excludeColumns?: string[];
        includeDeleted?: boolean;
        containsField?: string;
        equalityFilter?: any;
        criteria?: any[];
    }

    @Serenity.Decorators.registerEditor("CustomServiceLookupEditorBase")
    export class CustomServiceLookupEditorBase<TOptions extends ServiceLookupEditorOptions, TItem> extends Serenity.Select2Editor<TOptions, TItem> {

        constructor(input: JQuery, opt?: TOptions) {
            super(input, opt);
        }

        protected getDialogTypeKey() {
            var dialogTypeKey = super.getDialogTypeKey();
            if (dialogTypeKey)
                return dialogTypeKey;

            var service = this.getService();
            if (Q.startsWith(service, "~/Services/"))
                service = service.substr("~/Services/".length);

            if (service.split('/').length == 3)
                service = service.substr(0, service.lastIndexOf('/'));

            return service.replace("/", ".");
        }

        protected getService(): string {
            return this.options.service;
        }

        protected getServiceUrl() {
            var url = this.getService();
            if (url == null)
                throw new Error("ServiceLookupEditor requires 'service' option to be configured!");

            if (!Q.startsWith(url, "~") && !Q.startsWith(url, "/") && url.indexOf('://') < 0)
                url = "~/Services/" + url;

            if (Q.startsWith(url, "~"))
                url = Q.resolveUrl(url);

            return url;
        }

        protected getIncludeColumns() {
            var include = this.options.includeColumns || [];
            var idField = this.getIdField();

            if (idField && include.indexOf(idField) < 0)
                include.push(idField);

            var textField = this.getTextField();
            if (textField && include.indexOf(textField) < 0)
                include.push(textField);

            return include;
        }

        protected getSort() {
            return this.options.sort || (this.getTextField() ? [this.getTextField()] : null);
        }

        protected getCascadeCriteria(): any[] {

            var val = this.get_cascadeValue();

            if (val == null || val === '') {

                if (!Q.isEmptyOrNull(this.get_cascadeField())) {
                    return ['1', '=', '0'];
                }

                return null;
            }

            var fld = this.get_cascadeField();

            return [[fld], '=', val];
        }

        protected getFilterCriteria(): any[] {
            var val = this.get_filterValue();

            if (val == null || val === '') {
                return null;
            }

            var fld = this.get_filterField();
            return [[fld], '=', val];
        }

        protected getIdListCriteria(idList: any[]): any[] {
            if (idList == null)
                return null;

            if (idList.length == 0)
                return ['0', '=', '1'];

            var idField = this.getIdField();
            if (idField == null)
                throw new Error("ServiceLookupEditor requires 'idField' option to be configured!");

            return [[idField], 'in', [idList]];
        }

        protected getCriteria(query: Serenity.Select2SearchQuery): any[] {
            return Serenity.Criteria.and(
                Serenity.Criteria.and(this.getIdListCriteria(query.idList), this.options.criteria),
                Serenity.Criteria.and(this.getCascadeCriteria(), this.getFilterCriteria()));
        }

        protected getListRequest(query: Serenity.Select2SearchQuery): Serenity.ListRequest {

            var request: Serenity.ListRequest = {};

            if (query.searchTerm)
                request.ContainsText = query.searchTerm;

            request.Sort = this.getSort();
            request.ColumnSelection = this.options.columnSelection || Serenity.ColumnSelection.KeyOnly;
            request.IncludeColumns = this.getIncludeColumns();
            request.ExcludeColumns = this.options.excludeColumns;
            request.ContainsField = this.options.containsField;
            request.EqualityFilter = this.options.equalityFilter;
            request.Criteria = this.getCriteria(query);
            request.Skip = query.skip || 0;
            request.Take = query.take ? (query.checkMore ? query.take + 1 : query.take) : 0;
            request.IncludeDeleted = this.options.includeDeleted;
            request.ExcludeTotalCount = true;

            return request;
        }

        protected getServiceCallOptions(query: Serenity.Select2SearchQuery, results: (result: Serenity.Select2SearchResult<TItem>) => void): Serenity.ServiceOptions<Serenity.ListResponse<TItem>> {
            return {
                blockUI: false,
                url: this.getServiceUrl(),
                request: this.getListRequest(query),
                onSuccess: response => {
                    var items = response.Entities || [];

                    if (items && query.take && query.checkMore && response.Entities.length > items.length)
                        items = items.slice(0, query.take);

                    results({
                        items: items.slice(0, query.take),
                        more: query.checkMore && query.take && items.length > query.take
                    });
                }
            };
        }

        protected hasAsyncSource() {
            return true;
        }

        protected asyncSearch(query: Serenity.Select2SearchQuery, results: (result: Serenity.Select2SearchResult<TItem>) => void): Serenity.Select2SearchPromise {
            var opt = this.getServiceCallOptions(query, results);
            return Q.serviceCall(opt);
        }
    }

    @Serenity.Decorators.registerEditor('TestServiceLookupEditor')
    export class TestServiceLookupEditor extends CustomServiceLookupEditorBase<ServiceLookupEditorOptions, any> {
        constructor(hidden: JQuery, opt?: ServiceLookupEditorOptions) {
            super(hidden, opt);
        }
    }
}
```

Build, transform t4 template and using it in your Row.cs like this:

```csharp
[Common.Editors.TestServiceLookupEditor(Service = "Default/Supplier/List", IdField = "Id", TextField = "CompanyName", PageSize = 5)]
public Guid? SupplierId
{
    get { return Fields.SupplierId[this]; }
    set { Fields.SupplierId[this] = value; }
}
```

Good luck

**Reply by StevenKnoop** [reply]
Hi @minhhungit,

Thank you for the above. 
Do you perhaps know how to return and list the results that start with the given search term? Currently it is finding the results but it sorts it by alphabetical order so if I search "zad", I get a list of items that all contain "zad" in the name but the items that start with zad are at the bottom of the list. How do I show the items that start with the given search term first?

**Reply by waleed51** [reply]
@minhhungit  thank you for your efforts , i will try it and tell you

**Reply by minhhungit** [reply]
@StevenKnoop try to use `QuickSearch` with `SearchType.StartsWith`
```csharp
[QuickSearch(SearchType.StartsWith)]
public String CategoryName
{
    get => fields.CategoryName[this];
    set => fields.CategoryName[this] = value;
}
```

Btw, please create new discussion instead, it will be clear and better for tracking. Thanks

**Reply by waleed51** [reply]
@minhhungit  what should i replace "Default/Supplier/List" with

**Reply by minhhungit** [reply]
@waleed51 `Default/Supplier/List` is path of your service, for example in serenity demo project:

![image](https://user-images.githubusercontent.com/2279508/113494363-bba3f800-9511-11eb-8333-1f1facb9d8e2.png)

**Reply by waleed51** [reply]
thank you , but sorry for that it's not working yet  @minhhungit 
![image](https://user-images.githubusercontent.com/18492471/113501478-9d47f780-9525-11eb-87c4-c930806a6bb8.png)
![image](https://user-images.githubusercontent.com/18492471/113501491-b3ee4e80-9525-11eb-9689-79df68b09175.png)

**Reply by minhhungit** [reply]
Any error or something like that

**Reply by waleed51** [reply]
i think there is an error in 
![image](https://user-images.githubusercontent.com/18492471/113548535-fa0fe480-95ef-11eb-84af-6f08c3ac72e1.png)

but it's run and the lookup is empty 
@minhhungit

**Reply by minhhungit** [reply]
getIdField and getTextField is from Select2Editor, are you sure your class is inherited from `Serenity.Select2Editor<TOptions, TItem>`

**Reply by waleed51** [reply]
yes sure  @minhhungit 
![image](https://user-images.githubusercontent.com/18492471/113571325-7ae1d700-9616-11eb-924f-78170a634820.png)

**Reply by minhhungit** [reply]
It is hard for me to guess the problem without checking/debugging code.

Let's try to check if Serenity.Select2Editor in your project has those methods

**Reply by waleed51** [reply]
ok i will check and tell you @minhhungit
